src/Microsoft.AspNetCore.Routing/Matching/ILEmitTrieFactory.cs

```diff
@@ -0,0 +1,588 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#if IL_EMIT
+using System;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using System.Reflection.Emit;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.AspNetCore.Routing.Matching
+{
+    internal static class ILEmitTrieFactory
+    {
+        // The algorthm we use only works for ASCII text. If we find non-ASCII text in the input
+        // we need to reject it and let is be processed with a fallback technique.
+        public const int NotAscii = Int32.MinValue;
+
+        public static Func<string, int, int, int> Create(
+            int defaultDestination,
+            int exitDestination,
+            (string text, int destination)[] entries,
+            bool? vectorize)
+        {
+            var method = new DynamicMethod(
+                "GetDestination",
+                typeof(int),
+                new[] { typeof(string), typeof(int), typeof(int), });
+
+            GenerateMethodBody(method.GetILGenerator(), defaultDestination, exitDestination, entries, vectorize);
+
+#if IL_EMIT_SAVE_ASSEMBLY
+            SaveAssembly(method.GetILGenerator(), defaultDestination, exitDestination, entries, vectorize);
+#endif
+
+            return (Func<string, int, int, int>)method.CreateDelegate(typeof(Func<string, int, int, int>));
+        }
+
+        private static void GenerateMethodBody(
+            ILGenerator il,
+            int defaultDestination,
+            int exitDestination,
+            (string text, int destination)[] entries,
+            bool? vectorize)
+        {
+            // There's no value in vectorizing the computation if we're on 32bit or
+            // if no string is long enough. We do the vectorized comparison with uint64 ulongs
+            // which isn't beneficial if they don't map to the native size of the CPU. The
+            // vectorized algorithm introduces additional overhead for casing.
+            //
+            // Vectorize by default on 64bit (allow override for testing)
+            vectorize = vectorize ?? (IntPtr.Size == 8);
+
+            // Don't vectorize if all of the strings are small (prevents allocating unused locals)
+            vectorize &= entries.Any(e => e.text.Length >= 4);
```

Comment:
Since this doesn't effect the result, how about splitting the preceding logic out into a `ResolveIsVectorized` method and add some unit tests?

Reply:
Adding